lib/src/range/mod.rs

```diff
@@ -43,3 +46,82 @@ impl Hash for Semver {
         self.components.hash(state);
     }
 }
+
+impl fmt::Display for Range {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let str = self
+            .set
+            .iter()
+            .map(|v| match v {
+                Constraint::Any => "*".to_string(),
+                Constraint::Single(v) => format!("={}", v),
+                Constraint::Contiguous(v1, v2) => {
+                    if v2.major == v1.major + 1 && v2.minor == 0 && v2.patch == 0 {
+                        let v = chomp(v1);
+                        if v1.major == 0 {
+                            if v1.components.len() == 1 {
+                                "^0".to_string()
+                            } else {
+                                format!(">={}<1", v)
+                            }
+                        } else {
+                            format!("^{}", v)
+                        }
+                    } else if v2.major == v1.major && v2.minor == v1.minor + 1 && v2.patch == 0 {
+                        let v = chomp(v1);
+                        format!("~{}", v)
+                    } else if v2.major == usize::MAX {
+                        let v = chomp(v1);
+                        format!(">={}", v)
+                    } else if at(v1, v2) {
+                        format!("@{}", v1)
+                    } else {
+                        format!(">={}<{}", chomp(v1), chomp(v2))
+                    }
+                }
+            })
+            .collect::<Vec<_>>()
+            .join(",");
+        write!(f, "{}", str)
+    }
+}
+
+/// checks @ syntax, eg. node@22.1
+fn at(v1: &Semver, v2: &Semver) -> bool {
+    let mut cc1 = v1.components.clone();
+    let cc2 = &v2.components;
+
+    if cc1.len() > cc2.len() {
+        return false;
+    }
+
+    // Ensure cc1 and cc2 have the same length by appending 0s to cc1
+    while cc1.len() < cc2.len() {
+        cc1.push(0);
+    }
+
+    if last(&cc1) != last(cc2) - 1 {
+        return false;
+    }
```

Comment:
what is this doing? i'm getting underflows trying to test this with e.g. 1.0.0

Reply:
i've reversed the side that gets the +/-1. but what is this comparing?

Reply:
dunno this is a direct port of the ts. Let me noodle it.

Reply:
there might be a simpler way; `@` is really a `Range`? if so, we can just use `Range::parse()` and add `@` to the parser. i think it's `~`, right? even looking in `libpkgx` it's not precisely clear what that `- 1` is supposed to do.